### PR TITLE
Raw Meat overhaul

### PR DESCRIPTION
**Pull request by Marsunpaisti**
https://gfycat.com/PotableBlandCranefly

A complete overhaul of the current corpse system.
- Instead of corpses, "Raw Meat" items are spawned, which on pickup get transferred to a custom hidden inventory. Amount of meat carried is displayed on the Drop Meat ability.
- Raw Meat items decay like corpses using the ExpiringRawMeat package which removes them after being on ground for 60 seconds.

W3X Changes by jaccouille merged into this w3x:
- Several tooltip fixes
- Fish have Ghost(visible) to avoid blocking

### CODE REVIEW

File: wurst/__legacy__/trolls/gatherer/ItemWarp.jurst [hunk 1]
```diff
@@ -50,46 +50,49 @@ private function PullItem takes nothing returns nothing
...
-    call SetItemVisible(i, false)
-    
-    //call SetHandleInt(d.t,"struct",d)
-    call SaveInteger(udg_GameHash,GetHandleId(d.t),StringHash("struct"),d castTo int)   
-    call TimerStart(d.t, .03, true, function PullItem)
-    
-    set i = null
-    set u = null
-    set it = null
+    //local location loc = GetUnitLoc(u)
```
Comment (jonathanmclaus) [inline wurst/__legacy__/trolls/gatherer/ItemWarp.jurst]: Can you just add a short-circuit check here? Should reduce the diff greatly.

Comment (Marsunpaisti) [reply wurst/__legacy__/trolls/gatherer/ItemWarp.jurst]: Done

File: wurst/systems/meatsystem/MeatSystem.wurst [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    let target = GetOrderTargetItem()
+
+    //Smart pickup system -> simulate picking up meat like an item if full inventory, like in Bannars StackNSplit system
+    if (target != null
+        and target.isPickupable()
+        and target.getTypeId() == ITEM_RAW_MEAT
+        and u.isInventoryFull()
+        and (meatInventoryInstances.has(u) or isUnitMeatCarrier(u)))
+
+        if meatInventoryInstances.has(u)
```
Comment (jonathanmclaus) [inline wurst/systems/meatsystem/MeatSystem.wurst]: Create the `instance` right here and don't put the rest of the logic in a branch.

File: wurst/systems/meatsystem/MeatSystem.wurst [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        if meatInventoryInstances.has(caster) or isUnitMeatCarrier(caster)
+            unitTryPickupMeatArea(caster)
+
+    registerSpellEffectEvent(ABILITY_ID_PICKUPMEAT_TROLL) ->
+        let caster = GetSpellAbilityUnit()
+        if meatInventoryInstances.has(caster) or isUnitMeatCarrier(caster)
+            unitTryPickupMeatArea(caster)
+
+    //Special fix for beastmaster to get his abilities back after transforming
+    for s in ABILITY_IDS_SHAPESHIFTER
```
Comment (jonathanmclaus) [inline wurst/systems/meatsystem/MeatSystem.wurst]: Can you avoid doing this manually, counting the new system? If the ability is permanent it should transfer between transformation automatically.

File: wurst/systems/meatsystem/EatRawMeat.wurst [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+let eatRawMeatInstances = new HashMap<unit, EatRawMeatInstance>()
+
+class EatRawMeatInstance
+    unit caster
+    bool isChanneling
+
+    function healthLoop()
+        doPeriodicallyTimed(ANIMATION_PERIOD, CHANNEL_TIME) cb ->
+            if this.isChanneling and caster.isAlive()
```
[the remark this answers is not in the document]
Comment (Marsunpaisti) [reply wurst/systems/meatsystem/EatRawMeat.wurst]: changed

File: wurst/systems/meatsystem/MeatSystem.wurst [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+
+    construct(unit owner)
+        this.ownerUnit = owner
+        this.ownerPlayer = ownerUnit.getOwner()
+        this.carriedMeat = new LinkedList<item>()
+        this.isTrollInventory = UnitEntity.findForUnit(ownerUnit) instanceof Troll
+        updateMeatAbilities()
+
+    function emptyToGround()
+        let pos = ownerUnit.getPos()
```
[the remark this answers is not in the document]
Comment (Marsunpaisti) [reply wurst/systems/meatsystem/MeatSystem.wurst]: Fixed

File: wurst/systems/meatsystem/MeatSystem.wurst [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+        if carriedMeat.size() >= 0
+            carriedMeat.pop()
+            updateMeatAbilities()
+            return true
+        return false
+
+    function onOwnerDeath()
+        //Find an offset position for the meat so items are separated on drop
+        let pos = ownerUnit.getPos()
+        for angleIndex = 0 to 3
```
[the remark this answers is not in the document]
Comment (Marsunpaisti) [reply wurst/systems/meatsystem/MeatSystem.wurst]: changed

File: wurst/systems/meatsystem/MeatSystem.wurst [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+                break
+            else if angleIndex == 3
+                emptyToGround(pos)
+
+        destroy this
+
+    function canPickupMeat() returns bool
+        return carriedMeat.size() < MAX_MEAT_COUNT
+
+    function transferMeatInventoryToUnit(unit newOwner)
```
[the remark this answers is not in the document]
Comment (Marsunpaisti) [reply wurst/systems/meatsystem/MeatSystem.wurst]: Yeah it was supposed to change player too. I forgot that. Changed.

File: wurst/systems/meatsystem/MeatSystem.wurst [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+        //Remove old callback if it exists
+        if this.smartPickupCallback != null
+            destroy this.smartPickupCallback
+            this.smartPickupCallback = null
+
+        //Start polling for when we reach the pickup target
+        this.smartPickupCallback = doPeriodically(ANIMATION_PERIOD) cb ->
+            let collisionSize = ownerUnit.getCollisionSize()
+            let rangeSq = (SMART_PICKUP_RANGE + collisionSize) * (SMART_PICKUP_RANGE + collisionSize)
```
[the remark this answers is not in the document]
Comment (Marsunpaisti) [reply wurst/systems/meatsystem/MeatSystem.wurst]: Changed

File: wurst/systems/meatsystem/MeatSystem.wurst (hunk 2 above)
Comment (Marsunpaisti) [reply wurst/systems/meatsystem/MeatSystem.wurst]: Changed

File: wurst/systems/meatsystem/MeatSystem.wurst (hunk 3 above)
Comment (Marsunpaisti) [reply wurst/systems/meatsystem/MeatSystem.wurst]: WIth new BM system we dont need this.

File: wurst/systems/meatsystem/EatRawMeat.wurst [hunk 9]
```diff
@@ -70,10 +70,7 @@ class EatRawMeatInstance
         doPeriodicallyTimed(ANIMATION_PERIOD, CHANNEL_TIME) cb ->
             if this.isChanneling and caster.isAlive()
                 caster.setHP(caster.getHP() + HP_PER_SECOND * ANIMATION_PERIOD)
-            else
-                cb.stop()
-
-            if cb.isLast() or not (this.isChanneling and caster.isAlive())
+            else if cb.isLast() or not (this.isChanneling and caster.isAlive())
```
Comment (jonathanmclaus) [inline wurst/systems/meatsystem/EatRawMeat.wurst]: This would be the same as an `else` branch without a condition.

Comment (Marsunpaisti) [reply wurst/systems/meatsystem/EatRawMeat.wurst]: Oops. Looks like i rushed that a bit. Fixed

**Comment by jonathanmclaus** [comment]
A few improvements could be made:

Upgrading a unit to a unit that already has a meat inventory is the only time that `dropToGround` is called without a position, so that's probably not needed in the destructor of the instance. Death calls the destructor but empties the instance ahead of time.

Eating raw meat could probably use a dummy ability to spawn a corpse, add an ability based on eat raw, issue the order, and then remove the ability.

**Pull request merged**